### ISSUE TITLE
metaclass instance variable should be seen as class ClassVar

### ISSUE THREAD

**Issue by beauxq** [microsoft/pyright#7372]
**Describe the bug**
When a metaclass has an instance variable, the class (that is an instance of the metaclass) should have that variable as a `ClassVar`

The class is an instance of the metaclass.

**Code**

```python
from typing import Any
from typing_extensions import reveal_type


class M(type):
    """ ensure classes of this type have ClassVar `x` """
    x: list[int]

    def __new__(mcs, name: str, bases: tuple[type, ...], dct: dict[str, Any]) -> "M":
        # construct class
        new_class = super().__new__(mcs, name, bases, dct)
        if "x" not in dct:
            new_class.x = [2]
        return new_class


class C(metaclass=M):
    x = []  # mouseover says list[Unknown] when it should know list[int]


class D(metaclass=M):
    pass

    def f(self) -> None:
        self.x = [5]  # no type error reported
        # should report assigning instance variable over class variable


class E(metaclass=M):
    x = 3  # no type error reported - expected '"Literal[3]" is incompatible with "list[int]"'


print(C.x)  # []
# pyright reports 'Type of "x" is partially unknown' - it should know list[int]

print(D.x)  # [2]

print(E.x)  # 3

reveal_type(C.x)  # mypy: correctly sees list[int]  pyright: list[Unknown] should be list[int]

reveal_type(D.x)  # list[int]

reveal_type(E.x)  # mypy: list[int]  pyright: int
# I don't think it matters what this reports,
# because there should be a type error at assignment above.

```

**VS Code extension or command-line**
vscode pylance

**Comment by erictraut** [comment]
Some of this touches on areas of the typing spec that are underspecified.

I agree with most of the observations you've made in the code sample above. One that I don't agree with is the statement "should report assigning instance variable over class variable". This violates an [existing design decisions in pyright](https://microsoft.github.io/pyright/#/type-concepts-advanced?id=pure-class-variables). Pyright distinguishes between "pure class variables", "regular class variables" and "pure instance variables". I consider this a "regular class variable", and overwriting it with an instance variable is therefore allowed. 

Unless/until we have an agreed-upon clarification in the typing spec, I don't plan to revisit this design decision. That said, I would like to see us get clarity in the typing spec on the issue of class and instance variables. It's on a long list of issues that I've been maintaining, but it's not one of the top priorities at the moment, so it will need to wait.

I agree with the other issues that you've flagged in your code sample:
1. If the type of a metaclass instance variable is declared, this type should apply to a class variable or instance variable in a class that is created by the metaclass.
2. If the type of a metaclass instance variable is declared, an attempt to redeclare that type (incompatibly) within a class should be considered an error.

**Comment by erictraut** [comment]
> Is there (or will there be) any way to define the "pure class variable" of the class from the metaclass (to get that error of "x" cannot be assigned through a class instance because it is a ClassVar)

There's currently no mechanism in the type system. Defining instance variables in metaclasses is a pretty unusual thing — something that static type checkers have not traditionally supported (as you've found). You're several sigmas outside of the norm here. Are you sure this is how you want to model your code? I don't know exactly what you're trying to accomplish, but I would recommend looking for a more standard way of doing it if you want your code to work well with the static type system and existing type checkers.

**Comment by beauxq** [comment]
I'm trying to add typing to an established code base where this is part of the public API used by many. Changing the model would break backwards compatibility.

**Comment by erictraut** [comment]
This is included in pyright 1.1.353, which I just published.
